tortoise/ninja_tortoise.go

```diff
@@ -490,35 +499,21 @@ func (ni *NinjaTortoise) updatePatSupport(p votingPattern, bids []types.BlockID,
 	ni.TPatSupport[p][idx] = votingPattern{id: pid, LayerID: idx}
 }
 
-func initTallyToBase(tally map[votingPattern]map[types.BlockID]vec, base votingPattern, p votingPattern) {
-	tally[p] = make(map[types.BlockID]vec)
-	for k, v := range tally[base] {
-		tally[p][k] = v
+func (ni *NinjaTortoise) initTallyToBase(base votingPattern, p votingPattern, windowStart types.LayerID) {
+	mp := make(map[BlockIDLayerTuple]vec, len(ni.TTally[base]))
+	for b, v := range ni.TTally[base] {
+		if b.Layer() < windowStart {
```

Comment:
Something doesn't add up here. This condition seems to all be satisfied (except maybe for the very first layers), because `b.Layer() < p_base <= windowStart`.
More generally, understanding the leak cause, it seems that this entire function is unnecessary: we never need to actually copy p_base's opinion, as it is irreversible, and in addition we should have done SaveOpinion to p_base opinion by here.

Reply:
well your kinda right, i guess that now that we treat everything beneath pbase as irreversible and we already evicted all the "irreversible" blocks from the maps, we only need to consider blocks above pbase, need to remember that the case where windowstart > pbase is not handled.
but what about pbase - hdist we have explicit votes there, and i think we have to consider them

Reply:
maybe we should switch to 	
windowStart = Max(**ni.PBase.Layer()-ni.Hdis**t, newlyr.Index()-Window+1)